File: coderust/strings/reverse_sentence.rb
def reverse_sentence(a_string)
  reverse_chars(a_string, 0, a_string.length - 1)
  str_position = 0
  i = 0

  while i < a_string.length
    if a_string[i] == ' ' or i == a_string.length
      reverse_chars(a_string, str_position, i-1)
      str_position = i + 1
    end

    i += 1
  end

  return a_string
end

def reverse_chars(a_string, first, last)

  while first <= last
    temp = a_string[first]
    a_string[first] = a_string[last]
    a_string[last] = temp

    first += 1
    last -= 1
  end
end

puts reverse_sentence('I am going to school')

File: ik_problems/strings_arrays/reverse_sentence.py
def reverse_sentence(sentence):
    n = len(sentence)
    input = list(sentence)
    rev_sen = reverse(input, 0, n-1)

    i = 0
    position = 0

    while i <= n:
        # this line is very important. reverse the order and you'll get an IndexError
        if i == n or rev_sen[i] == ' ':
            reverse(rev_sen, position, i - 1)
            position = i + 1

        i += 1

    return ''.join(rev_sen)



def reverse(a_list, first, last):
    while first < last:
        temp = a_list[first]
        a_list[first] = a_list[last]
        a_list[last] = temp

        first += 1
        last -= 1

    return a_list

sen = 'boys love girls'
print reverse_sentence(sen)
